pallets/oracle-data-collection/src/lib.rs

```diff
@@ -0,0 +1,358 @@
+			keys_updated: u32,
+		},
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// The account who trigger the action is not a collection admin.
+		IsNotAdmin,
+
+		/// The key is not in the collection.
+		KeyNotInCollection,
+
+		/// Collection size reached
+		MaxCollectionSize,
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Update the feeders associated to an specific key.
+		///
+		/// NOTE: In the future this should be go though change guard.
+		/// By now we trust in the admin to choose honest feeders without
+		/// invertor supervision
+		#[pallet::weight(1_000_000)]
+		#[pallet::call_index(2)]
+		pub fn update_feeders(
+			origin: OriginFor<T>,
+			collection_id: T::CollectionId,
+			key: T::OracleKey,
+			feeders: BoundedVec<T::AccountId, T::MaxFeedersPerKey>,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+
+			ensure!(
+				T::IsAdmin::check((who, collection_id)),
+				Error::<T>::IsNotAdmin
+			);
+
+			Self::mutate_and_remove_if_clean(&collection_id, &key, |info| {
+				info.feeders = feeders.clone();
+				Ok(())
+			})?;
+
+			Self::deposit_event(Event::<T>::UpdatedFeeders {
+				collection_id: collection_id,
+				key: key,
+				feeders,
+			});
+
+			Ok(())
+		}
+
+		/// Update the collection, doing the aggregation for each key in the
+		/// process
+		#[pallet::weight(1_000_000)]
+		#[pallet::call_index(3)]
+		pub fn update_collection(
+			origin: OriginFor<T>,
+			collection_id: T::CollectionId,
+		) -> DispatchResult {
+			ensure_signed(origin)?;
+
+			let values = Keys::<T>::iter_key_prefix(&collection_id)
+				.map(|key| Self::get(&key, &collection_id).map(|value| (key, value)))
+				.collect::<Result<Vec<_>, _>>()?;
+
+			let len = values.len();
+
+			let collection = BoundedBTreeMap::try_from(BTreeMap::from_iter(values))
+				.map_err(|()| Error::<T>::MaxCollectionSize)?;
+
+			Collection::<T>::insert(collection_id, collection);
+
+			Self::deposit_event(Event::<T>::UpdatedCollection {
+				collection_id: collection_id,
+				keys_updated: len as u32,
+			});
+
+			Ok(())
+		}
+	}
+
+	impl<T: Config> DataRegistry<T::OracleKey, T::CollectionId> for Pallet<T> {
+		type Collection = CachedCollection<T>;
+		type Data = (T::OracleValue, T::Moment);
+		#[cfg(feature = "runtime-benchmarks")]
+		type MaxCollectionSize = T::MaxCollectionSize;
+
+		fn get(
+			key: &T::OracleKey,
+			collection_id: &T::CollectionId,
+		) -> Result<Self::Data, DispatchError> {
+			let key_info = Keys::<T>::get(collection_id, key);
+			let fed_values = key_info
+				.feeders
+				.into_iter()
+				.map(|feeder| T::OracleProvider::get(&(feeder, *collection_id), key))
+				.collect::<Result<Vec<_>, _>>()?;
+
+			let (mut values, mut timestamps): (Vec<_>, Vec<_>) = fed_values.into_iter().unzip();
+
+			let value = util::median(&mut values).ok_or(Error::<T>::KeyNotInCollection)?;
+			let timestamp = util::median(&mut timestamps).ok_or(Error::<T>::KeyNotInCollection)?;
+
+			Ok((*value, *timestamp))
```

Comment:
This can probably be moved to an `AggregationProvider` trait or similar to handle this in the runtime.

I still want to think more about using the median for the timestamp, too. I think could work but not 100% sure.

Reply:
The above means:

- If 50% of the feeders have set the correct value, we would aggregate with the correct value.
- If 50% of the feeders are updated, the price will be updated.

If we have less than 50% of well-behaved/ updated feeders, then the precision of the value and timestamp obtained will decrease.

I think we can leave under this scenario, and even more, taking into account that the admin is incentivized to select good/friend/owned feeders.

WDYT? @mustermeiszer @hieronx